Simplify examples

mikekryjak opened the pull request:
Resolves https://github.com/FusionDLS/FusionDLS/issues/42.

Our current examples are far too verbose because they were initially made to allow the reproduction of a specific work package for STEP last year. Keeping them updated while refactoring the code is not the best use of our time. I moved them to a separate folder called `deprecated` just so a user can look in there for inspiration if required. We can deal with them later and maybe make more examples.

Instead, we now have just two examples: `basic_use` and `geometry_manipulation` well as the notebook used to perform the analytical test that forms the basis of our CI. 

`geometry_manipulation` documents the functionality of `Profile` to help maintain its functionality for the refactor. In preparing this notebook I had to add several dosctrings and improve some functionality of `Profile`.

mikekryjak committed:
Merge branch 'clearer-fixed-qpllt' into update-examples

mikekryjak committed:
Change default Ttol to 1e-3
1e-2 gives visibly wonky results in example notebook

mikekryjak committed:
Deprecate old examples and rename analytic test example

mikekryjak committed:
Add new concise basic use example

mikekryjak committed:
Formatting

mikekryjak committed:
More formatting

mikekryjak committed:
Add Profile.plot()

mikekryjak committed:
Make scale_bxbt warning message consistent with scale_Lc

mikekryjak committed:
Merge branch 'morph-improvements' into update-examples

mikekryjak committed:
Clarify Profile functions not meant for user to touch

LiamPattinson commented:
One of the lint failures here is [PLR0904](https://docs.astral.sh/ruff/rules/too-many-public-methods/). I don't think it's a very useful one, so it's probably worth adding to `pyproject.toml` under:

```toml
[tool.ruff.lint]
# ...
ignore = [
  # ...
  "PLR0904",
]

mikekryjak committed:
Improve Profile plotting

mikekryjak committed:
Add a ValueError on a gotcha and improve Profile docstrings

mikekryjak committed:
Change profile morphing workflow
The controls for poloidal flux expansion control sit in Profile.recalculate_topology(), so it has to be exposed to the user, at least for now. I added some prints to make this clearer. In the future this should be refactored as it's unwieldy.

mikekryjak committed:
Formatting

mikekryjak committed:
Add functionality to Profile.plot()

mikekryjak committed:
Minor formatting tweaks to plot_control_points

mikekryjak committed:
Add a ValueError to catch a gotcha in shift_points

mikekryjak commented:
@LiamPattinson I am extending this PR to also provide an example of what `Profile` does. It feels like we need that to make sure the functionality is preserved during the refactor. I am having to do a few tweaks to `Profile` to achieve this.

mikekryjak committed:
Initial work on new geometry_manipulation example
Found a bug where too much connection length is added when shifting points

mikekryjak committed:
Formatting

LiamPattinson commented:
That's probably a good idea. I can convert some of the examples to tests during the merge.

mikekryjak committed:
Rename "xpos" to "posx" in offsets to reduce confusion

mikekryjak committed:
Enable profile RZ plotting with upstream included

mikekryjak committed:
Missed input flag for full RZ profile plot

mikekryjak committed:
More work on geometry_manipulation example
Profile was being offset back to front, making it seem like there was a bug.

mikekryjak committed:
Merge branch 'morph-improvements' into update-examples

mikekryjak committed:
Fix rogue remaining "xpos"

mikekryjak committed:
Add content to geometry_manipulation notebook

mikekryjak committed:
Fix small typos

mikekryjak committed:
Add Profile.plot(mode="Bpol")

mikekryjak commented:
@LiamPattinson I just got Ruff complaining about [PLR6201](https://docs.astral.sh/ruff/rules/literal-membership/). I don't know if I agree with this one, and it's marked as unstable. Can we disable it?

I'm also getting an error for [I001](https://docs.astral.sh/ruff/rules/unsorted-imports/). I think this is an unfair one, because Ruff doesn't seem to be able to automatically sort the imports for me, and the website doesn't provide the rules to do so. That would make it pretty annoying for new developers.

LiamPattinson commented:
You should be able to fix any `I` lints just by running with `--fix`:

```bash
ruff check --select=I --fix
# Or to fix any fixable issues...
ruff check --fix
```

You can also do it using `isort`, from which Ruff derives its import sorting rules:

```bash
isort
```

`isort` is already configured in the `pyproject.toml`, so the two should be interchangeable. Import sorting is something I think we should keep, as it makes it much easier to figure out what's imported at a glance.

---

As for `PLR6201`, it's a rule that's only available in Ruff's preview mode, but it's also a safe fix, so you can fix it up with:

```bash
ruff check --preview --fix
```

I'd personally argue against the use of preview features in the CI though, so I think the better option is to remove `--preview` from the `ruff` call in `.github/workflows/lint.yml`.

mikekryjak committed:
Disable preview Ruff features
A test was failing on a preview rule - decided to disable them since having any linting at all is already a big step.

mikekryjak committed:
Sort imports

mikekryjak committed:
Merge branch 'main' into update-examples

mikekryjak committed:
Sort imports in basic_use.ipynb

mikekryjak commented:
Thanks @LiamPattinson, ready for review.

LiamPattinson commented on src/fusiondls/Profile.py:
```diff
@@ -200,6 +200,17 @@ def offset_control_points(self, offsets, factor=1):
+
+        Parameters
+        ----------
+        offsets : list of dictionaries
```
We're using `sphinx_autodoc_typehints` to generate our documentation, so rather that adding types here it would be preferable to add this as a type hint to the function argument, e.g. `offsets: list[dict]`. Similar changes could be made to a bunch of functions here.

mikekryjak replied:
Do you think you would be able to convert them?

LiamPattinson replied:
Sure, can do

LiamPattinson commented on src/fusiondls/Profile.py:
```diff
@@ -373,6 +393,89 @@ def plot_topology(self):
+        Parameters
+        ----------
+        mode : str
+            What to plot:
+                "Btot" - total B profile
+                "RZ" - RZ space leg profile (excl. above X-point)
+                "Spar_Spol" - Parallel vs poloidal connection length
```
I think this will render strangely in the html docs. This would do what you'd expect:

```suggestion
            What to plot:
            - "Btot": total B profile
            - "RZ": RZ space leg profile (excl. above X-point)
            - "Spar_Spol": Parallel vs poloidal connection length
```

LiamPattinson reviewed (approved):
All looks good to me! The only things that jumped out to me are minor issues related to docs and type-hints, but they're super-low priority and I'd be happy to merge without the changes.

mikekryjak committed:
Improve dosctring formatting in Profile.py

Co-authored-by: LiamPattinson <LiamPattinson@users.noreply.github.com>

LiamPattinson committed:
Update docstrings/typehints

Also formats Python code in notebooks and fixes up
a couple of Ruff warnings.

LiamPattinson committed:
Formatting fix

LiamPattinson commented:
Seems I found a rare divergence between `black` and `ruff format` in my last commit. I've updated docs and type hints to make everything read better in the HTML docs
